### ISSUE TITLE
BasicTimeSeries API changes

### ISSUE THREAD

**Issue by pavlis** [wangyinz/mspass#60]
In our call today we discussed the need to change the API for TimeSeries and Seismogram objects to allow them to work more cleanly with python through pybind11.   The issue to be addressed is the boundary between Metadata and data.   Some things now defined as class attributes in BasicTimeSeries can get out of sync with Metadata.   I think I have a good solution with virtual functions and getters and putters.   

To review the current BasicTimeSeries has several public attributes that are the problem.   Here they are with the documentation comment lines included:
```
/*!
Boolean defining if a data object has valid data or is to be ignored.
Data processing often requires data to be marked bad but keep the original
data around in case an error was made.  This boolean allows this capability.
**/
        bool live;
/*!
Sample interval.
**/
        double dt;
/*!
Data start time.  That is the time of the first sample of data.
**/
        double t0;
/*!
Number of data samples in this data object.
**/
        int ns;
/*!
Time reference standard for this data object.  Defined by enum Time_Reference
this currently is only one of two things.  When set as "UTC" the time
standard is an epoch time.  When set as "relative" time has no relationship
to any external standard but are relative to some arbitrary reference that must
ascertained by the algorithm by some other means (in seispp this is normally
done through a metadata object).  A classic example is multichannel data where
channels have a time relative to a shot time.
**/
        TimeReferenceType tref;
```
I suggest we change all these to pure getters and add some putters that are virtual.  In particular sans comments I recommend this:
```
...
template <class TimeType> class BasicTimeSeries
{
public:
  bool live();
  bool dead();  
  double dt();
  int npts();  // replacement for ns that is confusing anyway
  virtual TimeType t0()=0;
  void kill();  //sets live False
  void mark_live();   //sets live True
  virtual void set_dt(double sample_interval)=0;
  virtual void set_npts(int number_of_data_points)=0;
// other methods from BasicTimeSeries here (e.g. rtoa)
private:
  bool is_live;
  double sample_interval;
  int number_samples;
  TimeType ReferenceTime;
  TimeReferencetype tref;   //Instantiation of TimeType should set tref
};
```
CoreTimeSeries and CoreSeismogram would need to define the virtual methods.  The setters can guarantee the Metadata and internal attributes are consistent by having calls to the appropriate Metadata put methods.  

I may think better of the template to define TimeType as it might complicate things a bit, but I don't think it is too big a deal.  For mspass implementation I propose using the epoch time convention.   We can probably build a small python wrapper to return an obspy utctime object for the python bindings.  I don't think that would be hard.

**Comment by pavlis** [comment]
For the record, during our call we concluded that having TimeType as a template parameter for BasicTimeSeries was an unnecessary complication.  We'll stick with the original dogmatic approach of having time be an epoch time.  The  method no longer needs to be virtual:
```
double t0();
```
and be defined as returning an epoch time.   Children may choose to overload or define a different method name to return something else (e.g. a utctime() method).  

Emphasize that this interface will require CoreTimeSeries and CoreSeismogram to implement the virtual putters.   We will also need something missed:
```
virtual void set_t0(double t0);
```
This is virtual like the others because it needs to sync metadata.  

We talked about the alias problem with these attributes in Metadata.   Realized we could handle that in a couple of different ways:

1. We could hard code a set of alternate aliases.   This would complicate the putters as they would need to test for the existence of any of the aliases and automatically handle it some way.   Simplest would be a "if defined already" rule.   That is, if for example "npts" was not defined but "ns"  or "nsamp" were defined, set_npts could tests a list of valid aliases and change all of them.   If none were defined it would add the required key:value pair (e.g. "npts":100).   
2.  The constructors could define aliases through some mechanism with a default that would be backward compatible.   I think this is an unnecessary complexity and would complicate backward compatibility.  
3.  There may be a hybrid where the valid aliases for each attribute could be set through a set of special methods.   Seems something easier to add later if needed.   

Obviously I favor 1.   It would be a minor addition that would make this more solid.  Let me know your opinions and if you concur I can start implementing this.   I don' think it will be that hard for ccore.

**Comment by wangyinz** [comment]
I agree. We can go fancy later when necessary. btw, I think instead of naming the setters as `void set_t0(double t0)`, we should probably just overload it as `void t0(double t0)`,

**Comment by pavlis** [comment]
Ok, starting to implement this then.   

I'm not going to follow your suggestion to overload methods like t0.   The reason is set_to needs to be virtual and that will as a minimum create confusion and worse may not even be allowed.   I cannot see how the RTTI system could sort out overloaded methods with some virtual and some not.   Their use in this case would be totally opposite and even more likely to not be understood by the compiler.

**Comment by pavlis** [comment]
In working through this I realized the conversion would be simpler and much less error prone if the formerly public members dt, t0, ns, live, and tref were protected instead of private.   That way the children can access them directly and most instances of the C++ code require no changes.  The only exception is when one of the quantities is on the right hand side the context has to be established.  I am going through the code to make sure the context won't cause problems if the putters aren't used in preference to setting the raw attributes.

**Comment by pavlis** [comment]
I implemented a set_npts method for CoreSeismogram and CoreTimeSeries that does something that may be confusing and error prone.   I would like your thoughts on this, Ian.  

Reminder that the goal is to make sure that all elements of the object related to the data size (ns and metadata attributes) are in sync and especially in sync with the data.   This particular element of the basic attributes is the hardest to deal with because it links to that issue.   I particularly need your feedback because there are possible collision with the new python interface features you added we should consider now while this is active - i.e. not later when it turns up as a bug.  

What I implemented for both CoreTimeSeries and CoreSeismogram is this algorithm in pseudocode:
```
1)  Set the internal attribute (still called ns)
2)  set the unique name Metadata element (key npts)
3)  loop over all aliases and set them to this value too.
4) create or resize the data buffer for ns samples (different operations for std::vector and dmatrix)
5) initialize data buffer to zero
```
The problem is that algorithm assumes this method is being called during initialization and just initializing the buffer is appropriate.   The potential problem is a naive user could call the set_npts  method and find their data just vanished and be confused.   This would be an unusual error for an experienced programmer, but for students it could happen way way too easily.   

Problem defined I see three options:
1.  Change the algorithm and (optionally) the api a bit.  In particular, I suggest 
```
   int set_npts(const size_t number_samples);
```
where the int is used as an error return.  The algorithm would do this test:
```
1)if(number_samples<ns)
   i) copy data buffer
   ii) create new data buffer of length number_samples
   iii) copy truncated data to buffer
else if(number_samples>ns)
  i) copy data buffer
  ii) create new data buffer
  iii) copy old data into left side of data buffer
  iv) zero the extra space created with changing ns
else
  i) no else actually - do nothing if the size doesn't change
2) same algorithm as above for changed ns and metadata
```
The method would return 0 if the buffer wasn't changed, -1 if it was truncated, and +1 if it was zero padded.  

2.  We could make a new method that implements the above algorithm and depend upon the documentation. 

3.  We could consider the above algorithm dangerous and not implement it at all BUT add a method with this signature:
```
int sync_npts();
```
This method would use the buffer size to set ns and the associated metadata as you did in your modifications I saw.  In fact, that need is what drove me to write this comment.

I suggest we implement item 2 or 3.   I think under the hood, automatic messing around with the buffer size like that is dangerous.

**Comment by wangyinz** [comment]
Yeah, after thinking more about it, I agree that implicitly truncating data is not a good practice. My first thought to solving this problem is to make `npts` read-only just like the `size()` method in `std::vector`. Users have to explicitly call `resize()` to manipulate the size. Then, I realize ObsPy can face the same issue, so took a look at their implementation. It is actually similar to your item 3. Basically, they have a `Trace.count()` method that behaves like a wrapper around the read-only `numpy.array` attribute `Trace.data.size`, so it will always return the correct size. Then, there is the metadata key `npts` that can be accessed/changed through `Trace.stats.npts`, and it is weakly linked to the actual size of the internal array such that certain operations (e.g. re-assigning the `Trace.data`) will update this field while others (e.g. calling `Trace.data.resize()`) won't. Then, they have a [`Trace.verify()`](https://docs.obspy.org/packages/autogen/obspy.core.trace.Trace.verify.html#obspy.core.trace.Trace.verify) method that can reveal such potential inconsistency.

**Comment by pavlis** [comment]
I think a simple way to do that is to just not build a wrapper for set_npts at all.  That will make npts read only to the python user.   That behavior is useful in the C++ world - I just found a usage in the procedure ArrivalTimeReference where it is the right model.  I think adding a public sync_npts could be useful in multiple contexts.   We may want to hide it in the python wrappers by making it available but with the wrapper named to not use it lightly by giving having the wrapped name be _sync_npts or __sync_npts.

**Comment by pavlis** [comment]
Back to working on deconvolution code this morning and discovered there were clearly some hidden bugs created by the changes in behavior of set_npts compared to simply setting the attribute ns in the old api.   Had to make changes in multiple places in the deconvolution code and in several places in the code in algorithms.   There were some minor changes to CoreSeismogram and CoreTimeSeries that I did purely for efficiency.   

Although that was a pain to fix, it is the right behavior that method should define.   Any other behavior is dangerous.   It can create bugs that are mysterious as the order of defining things matter.  i.e. set_npts must be called before setting any data values in TimeSeries or Seismogram data vectors.  A potential bug in C++ code that won't happen in python code is using the push_back method incorrectly after calls to npts.   Since set_npts initialize the data vector to zeros that will produce unexpected results unless you call the std::vector clear method first.   That, however, is a common issue in std::vector algorithms so is an issue only in code I might try to convert as this evolves.

**Comment by wangyinz** [comment]
Yes, I noticed that issue. We will just live with it I guess.

**Comment by pavlis** [comment]
In working on the basic graphics code I uncovered the need for some basic amplitude scaling functions.  In starting to write those in C++ I discovered a glaring hole in the ccore API - neither TimeSeries nor Seismogram objects have some basic math operators.  Notably, we might want:
```
operator*=(const double);
operator-=(const double);
operator/=(const double);
operator+-(const double);
template <typename T> operator*=(const T&);
template <typename T> operator/=(const T&);
template <typename T> operator+=(const T&);
template <typename T> operator-=(const T&);
```
and the closely related binary operators like operator+, operator-, etc.   The templates are not literally what it would take this implement that as they couldn't be implemented without specializations to each data type that make the template mostly baggage. 

I bring this up as an issue as it isn't clear that creating these operators is a good idea.   The reason is it will break our ideas of ProcessingHistory.   All of the above will alter the data so technically should have the option (at least) of creating a history record.   Putting the history coding into the operators wouldn't work as I don't see how to overload the operator methods and make the history preservation optional.   

Right now I'm working on a scaling function, which is really no different than calls to operator*=.   I don't need operator*= to implement this feature, but it brings this topic up for discussion.   Given the problem I raised above I suspect we should not implement the math operators on TimeSeries and Seismogram until we find a more glaring need for them.

**Comment by wangyinz** [comment]
That's a very good point. I can totally see the need for such operators as they are the most intuitive to use in some cases, so the real problem is in the history mechanism. Maybe one way to resolve it is to have a separate variable that controls whether to enable history mechanism in each object, so we have:
```
class  BasicProcessingHistory
{
  public:
    void enable_history();
    void disable_history();
  protected:
    bool enabled=false;
}
```
Then we can check `enabled` to decide whether to retain history for the operators or not. If we do this, I guess we will need to make sure all other functions follows the same behavior. Also, I am not sure what `disable_history()` should behave. Should it remove the existing history records? Probably not, but then we can end up with something confusing with only partial history. Maybe we don't want the history to be disabled once enabled.

**Comment by wangyinz** [comment]
Closing this as completed.

`BasicTimeSeries` now exposes the live/dead state and timing/sample-count values through the API described here (`live`, `dead`, `kill`, `npts`, `set_dt`, and `set_npts`), and `TimeSeries`/`Seismogram` use the resulting model. The original public-field synchronization problem has therefore been resolved.

A desired arithmetic operator or history interaction should be proposed separately against the current classes.
